roachtest: tpchvec failed

[(roachtest).tpchvec failed](https://teamcity.cockroachdb.com/viewLog.html?buildId=1780216&tab=buildLog) on [master@5f9a71adb995837bcff27b9456188018434be4b8](https://github.com/cockroachdb/cockroach/commits/5f9a71adb995837bcff27b9456188018434be4b8):

```
The test failed on branch=master, cloud=gce:
test artifacts and logs in: /home/agent/work/.go/src/github.com/cockroachdb/cockroach/artifacts/20200303-1780216/tpchvec/run_1
	tpchvec.go:531,test_runner.go:741: error with attached stack trace:
		    main.execCmdWithBuffer
		    	/home/agent/work/.go/src/github.com/cockroachdb/cockroach/pkg/cmd/roachtest/cluster.go:426
		    main.(*cluster).RunWithBuffer
		    	/home/agent/work/.go/src/github.com/cockroachdb/cockroach/pkg/cmd/roachtest/cluster.go:2031
		    main.registerTPCHVec.func1
		    	/home/agent/work/.go/src/github.com/cockroachdb/cockroach/pkg/cmd/roachtest/tpchvec.go:520
		    main.(*testRunner).runTest.func2
		    	/home/agent/work/.go/src/github.com/cockroachdb/cockroach/pkg/cmd/roachtest/test_runner.go:741
		    runtime.goexit
		    	/usr/local/go/src/runtime/asm_amd64.s:1357
		  - error with embedded safe details: %s
		    -- arg 1: <string>
		  - /home/agent/work/.go/src/github.com/cockroachdb/cockroach/bin/roachprod run teamcity-1780216-1583223197-18-n3cpu4:1 -- ./workload run tpch --concurrency=1 --db=tpch --max-ops=3 --queries=4 --vectorize=experimental_on {pgurl:1-3}:
		  - exit status 1

	cluster.go:1410,context.go:135,cluster.go:1399,test_runner.go:778: dead node detection: /home/agent/work/.go/src/github.com/cockroachdb/cockroach/bin/roachprod monitor teamcity-1780216-1583223197-18-n3cpu4 --oneshot --ignore-empty-nodes: exit status 1 2: 13382
		1: 15347
		3: dead
		Error:  3: dead

```

<details><summary>More</summary><p>


Artifacts: [/tpchvec](https://teamcity.cockroachdb.com/viewLog.html?buildId=1780216&tab=artifacts#/tpchvec)

[See this test on roachdash](https://roachdash.crdb.dev/?filter=status%3Aopen+t%3A.%2Atpchvec.%2A&sort=title&restgroup=false&display=lastcommented+project)
<sub>powered by [pkg/cmd/internal/issues](https://github.com/cockroachdb/cockroach/tree/master/pkg/cmd/internal/issues)</sub></p></details>

[(roachtest).tpchvec failed](https://teamcity.cockroachdb.com/viewLog.html?buildId=1786843&tab=buildLog) on [master@c9b189b3a2171c8da78c75fe80dd02933e18a5e5](https://github.com/cockroachdb/cockroach/commits/c9b189b3a2171c8da78c75fe80dd02933e18a5e5):

```
The test failed on branch=master, cloud=gce:
test artifacts and logs in: /home/agent/work/.go/src/github.com/cockroachdb/cockroach/artifacts/20200305-1786843/tpchvec/run_1
	tpchvec.go:531,test_runner.go:741: error with attached stack trace:
		    main.execCmdWithBuffer
		    	/home/agent/work/.go/src/github.com/cockroachdb/cockroach/pkg/cmd/roachtest/cluster.go:426
		    main.(*cluster).RunWithBuffer
		    	/home/agent/work/.go/src/github.com/cockroachdb/cockroach/pkg/cmd/roachtest/cluster.go:2031
		    main.registerTPCHVec.func1
		    	/home/agent/work/.go/src/github.com/cockroachdb/cockroach/pkg/cmd/roachtest/tpchvec.go:520
		    main.(*testRunner).runTest.func2
		    	/home/agent/work/.go/src/github.com/cockroachdb/cockroach/pkg/cmd/roachtest/test_runner.go:741
		    runtime.goexit
		    	/usr/local/go/src/runtime/asm_amd64.s:1357
		  - error with embedded safe details: %s
		    -- arg 1: <string>
		  - /home/agent/work/.go/src/github.com/cockroachdb/cockroach/bin/roachprod run teamcity-1786843-1583394619-19-n3cpu4:1 -- ./workload run tpch --concurrency=1 --db=tpch --max-ops=3 --queries=4 --vectorize=experimental_on {pgurl:1-3}:
		  - exit status 1

```

<details><summary>More</summary><p>


Artifacts: [/tpchvec](https://teamcity.cockroachdb.com/viewLog.html?buildId=1786843&tab=artifacts#/tpchvec)

[See this test on roachdash](https://roachdash.crdb.dev/?filter=status%3Aopen+t%3A.%2Atpchvec.%2A&sort=title&restgroup=false&display=lastcommented+project)
<sub>powered by [pkg/cmd/internal/issues](https://github.com/cockroachdb/cockroach/tree/master/pkg/cmd/internal/issues)</sub></p></details>

[(roachtest).tpchvec failed](https://teamcity.cockroachdb.com/viewLog.html?buildId=1796240&tab=buildLog) on [master@c473f40078994551cebcbe00fdbf1fa388957658](https://github.com/cockroachdb/cockroach/commits/c473f40078994551cebcbe00fdbf1fa388957658):

```
The test failed on branch=master, cloud=gce:
test artifacts and logs in: /home/agent/work/.go/src/github.com/cockroachdb/cockroach/artifacts/20200309-1796240/tpchvec/run_1
	tpchvec.go:531,test_runner.go:741: error with attached stack trace:
		    main.execCmdWithBuffer
		    	/home/agent/work/.go/src/github.com/cockroachdb/cockroach/pkg/cmd/roachtest/cluster.go:426
		    main.(*cluster).RunWithBuffer
		    	/home/agent/work/.go/src/github.com/cockroachdb/cockroach/pkg/cmd/roachtest/cluster.go:2031
		    main.registerTPCHVec.func1
		    	/home/agent/work/.go/src/github.com/cockroachdb/cockroach/pkg/cmd/roachtest/tpchvec.go:520
		    main.(*testRunner).runTest.func2
		    	/home/agent/work/.go/src/github.com/cockroachdb/cockroach/pkg/cmd/roachtest/test_runner.go:741
		    runtime.goexit
		    	/usr/local/go/src/runtime/asm_amd64.s:1357
		  - error with embedded safe details: %s
		    -- arg 1: <string>
		  - /home/agent/work/.go/src/github.com/cockroachdb/cockroach/bin/roachprod run teamcity-1796240-1583738442-18-n3cpu4:1 -- ./workload run tpch --concurrency=1 --db=tpch --max-ops=3 --queries=4 --vectorize=experimental_on {pgurl:1-3}:
		  - exit status 1

```

<details><summary>More</summary><p>


Artifacts: [/tpchvec](https://teamcity.cockroachdb.com/viewLog.html?buildId=1796240&tab=artifacts#/tpchvec)

[See this test on roachdash](https://roachdash.crdb.dev/?filter=status%3Aopen+t%3A.%2Atpchvec.%2A&sort=title&restgroup=false&display=lastcommented+project)
<sub>powered by [pkg/cmd/internal/issues](https://github.com/cockroachdb/cockroach/tree/master/pkg/cmd/internal/issues)</sub></p></details>

[(roachtest).tpchvec failed](https://teamcity.cockroachdb.com/viewLog.html?buildId=1804460&tab=buildLog) on [master@793a9200c16693aff32aa6a4dd9d8bbcbddb30aa](https://github.com/cockroachdb/cockroach/commits/793a9200c16693aff32aa6a4dd9d8bbcbddb30aa):

```
The test failed on branch=master, cloud=gce:
test artifacts and logs in: /home/agent/work/.go/src/github.com/cockroachdb/cockroach/artifacts/20200312-1804460/tpchvec/run_1
	tpchvec.go:531,test_runner.go:747: error with attached stack trace:
		    main.execCmdWithBuffer
		    	/home/agent/work/.go/src/github.com/cockroachdb/cockroach/pkg/cmd/roachtest/cluster.go:426
		    main.(*cluster).RunWithBuffer
		    	/home/agent/work/.go/src/github.com/cockroachdb/cockroach/pkg/cmd/roachtest/cluster.go:2031
		    main.registerTPCHVec.func1
		    	/home/agent/work/.go/src/github.com/cockroachdb/cockroach/pkg/cmd/roachtest/tpchvec.go:520
		    main.(*testRunner).runTest.func2
		    	/home/agent/work/.go/src/github.com/cockroachdb/cockroach/pkg/cmd/roachtest/test_runner.go:747
		    runtime.goexit
		    	/usr/local/go/src/runtime/asm_amd64.s:1357
		  - error with embedded safe details: %s
		    -- arg 1: <string>
		  - /home/agent/work/.go/src/github.com/cockroachdb/cockroach/bin/roachprod run teamcity-1804460-1583998462-19-n3cpu4:1 -- ./workload run tpch --concurrency=1 --db=tpch --max-ops=3 --queries=4 --vectorize=experimental_on {pgurl:1-3}:
		  - exit status 1

```

<details><summary>More</summary><p>


Artifacts: [/tpchvec](https://teamcity.cockroachdb.com/viewLog.html?buildId=1804460&tab=artifacts#/tpchvec)

[See this test on roachdash](https://roachdash.crdb.dev/?filter=status%3Aopen+t%3A.%2Atpchvec.%2A&sort=title&restgroup=false&display=lastcommented+project)
<sub>powered by [pkg/cmd/internal/issues](https://github.com/cockroachdb/cockroach/tree/master/pkg/cmd/internal/issues)</sub></p></details>

--- a/pkg/sql/colexec/hashjoiner.go
+++ b/pkg/sql/colexec/hashjoiner.go
@@ -59,10 +59,6 @@ type hashJoinerSourceSpec struct {
 	// hash join.
 	eqCols []uint32
 
-	// outCols specify the indices of the columns that should be outputted by the
-	// hash joiner.
-	outCols []uint32
-
 	// sourceTypes specify the types of the input columns of the source table for
 	// the hash joiner.
 	sourceTypes []coltypes.T
@@ -227,7 +223,6 @@ func (hj *hashJoiner) Init() {
 		hashTableNumBuckets,
 		hj.spec.right.sourceTypes,
 		hj.spec.right.eqCols,
-		hj.spec.right.outCols,
 		false, /* allowNullEquality */
 		hashTableFullMode,
 	)
@@ -282,9 +277,12 @@ func (hj *hashJoiner) build(ctx context.Context) {
 	hj.state = hjProbing
 }
 
+// emitUnmatched populates the output batch to emit tuples from the build side
+// that didn't get a match. This will be called only for RIGHT OUTER and FULL
+// OUTER joins.
 func (hj *hashJoiner) emitUnmatched() {
 	// Set all elements in the probe columns of the output batch to null.
-	for i := range hj.spec.left.outCols {
+	for i := range hj.spec.left.sourceTypes {
 		outCol := hj.output.ColVec(i)
 		outCol.Nulls().SetNulls()
 	}
@@ -299,11 +297,10 @@ func (hj *hashJoiner) emitUnmatched() {
 		hj.emittingUnmatchedState.rowIdx++
 	}
 
-	outCols := hj.output.ColVecs()[len(hj.spec.left.outCols) : len(hj.spec.left.outCols)+len(hj.ht.outCols)]
-	for outColIdx, inColIdx := range hj.ht.outCols {
-		outCol := outCols[outColIdx]
-		valCol := hj.ht.vals.ColVec(int(inColIdx))
-		colType := hj.ht.valTypes[inColIdx]
+	outCols := hj.output.ColVecs()[len(hj.spec.left.sourceTypes) : len(hj.spec.left.sourceTypes)+len(hj.spec.right.sourceTypes)]
+	for i, colType := range hj.spec.right.sourceTypes {
+		outCol := outCols[i]
+		valCol := hj.ht.vals.ColVec(i)
 		// NOTE: this Copy is not accounted for because we don't want for memory
 		// limit error to occur at this point - we have already built the hash
 		// table and now are only consuming the left source one batch at a time,
@@ -454,49 +451,48 @@ func (hj *hashJoiner) congregate(nResults int, batch coldata.Batch, batchSize in
 	// concern here is only for the variable-sized types that exceed our
 	// estimations.
 
-	rightColOffset := len(hj.spec.left.outCols)
-	// If the hash table is empty, then there is nothing to copy. The nulls
-	// will be set below.
-	if hj.ht.vals.Length() > 0 {
-		outCols := hj.output.ColVecs()[rightColOffset : rightColOffset+len(hj.ht.outCols)]
-		for outColIdx, inColIdx := range hj.ht.outCols {
-			outCol := outCols[outColIdx]
-			valCol := hj.ht.vals.ColVec(int(inColIdx))
-			colType := hj.ht.valTypes[inColIdx]
-			// Note that if for some index i, probeRowUnmatched[i] is true, then
-			// hj.buildIdx[i] == 0 which will copy the garbage zeroth row of the
-			// hash table, but we will set the NULL value below.
-			outCol.Copy(
-				coldata.CopySliceArgs{
-					SliceArgs: coldata.SliceArgs{
-						ColType:   colType,
-						Src:       valCol,
-						SrcEndIdx: nResults,
-						Sel:       hj.probeState.buildIdx,
+	if hj.spec.joinType != sqlbase.LeftSemiJoin && hj.spec.joinType != sqlbase.LeftAntiJoin {
+		rightColOffset := len(hj.spec.left.sourceTypes)
+		// If the hash table is empty, then there is nothing to copy. The nulls
+		// will be set below.
+		if hj.ht.vals.Length() > 0 {
+			outCols := hj.output.ColVecs()[rightColOffset : rightColOffset+len(hj.spec.right.sourceTypes)]
+			for i, colType := range hj.spec.right.sourceTypes {
+				outCol := outCols[i]
+				valCol := hj.ht.vals.ColVec(i)
+				// Note that if for some index i, probeRowUnmatched[i] is true, then
+				// hj.buildIdx[i] == 0 which will copy the garbage zeroth row of the
+				// hash table, but we will set the NULL value below.
+				outCol.Copy(
+					coldata.CopySliceArgs{
+						SliceArgs: coldata.SliceArgs{
+							ColType:   colType,
+							Src:       valCol,
+							SrcEndIdx: nResults,
+							Sel:       hj.probeState.buildIdx,
+						},
 					},
-				},
-			)
+				)
+			}
 		}
-	}
-	if hj.spec.left.outer {
-		// Add in the nulls we needed to set for the outer join.
-		for outColIdx := range hj.ht.outCols {
-			outCol := hj.output.ColVec(outColIdx + rightColOffset)
-			nulls := outCol.Nulls()
-			for i, isNull := range hj.probeState.probeRowUnmatched {
-				if isNull {
-					nulls.SetNull(i)
+		if hj.spec.left.outer {
+			// Add in the nulls we needed to set for the outer join.
+			for i := range hj.spec.right.sourceTypes {
+				outCol := hj.output.ColVec(i + rightColOffset)
+				nulls := outCol.Nulls()
+				for i, isNull := range hj.probeState.probeRowUnmatched {
+					if isNull {
+						nulls.SetNull(i)
+					}
 				}
 			}
 		}
 	}
 
-	outCols := hj.output.ColVecs()[:len(hj.spec.left.outCols)]
-	for outColIdx, inColIdx := range hj.spec.left.outCols {
-		outCol := outCols[outColIdx]
-		valCol := batch.ColVec(int(inColIdx))
-		colType := hj.spec.left.sourceTypes[inColIdx]
-
+	outCols := hj.output.ColVecs()[:len(hj.spec.left.sourceTypes)]
+	for i, colType := range hj.spec.left.sourceTypes {
+		outCol := outCols[i]
+		valCol := batch.ColVec(i)
 		outCol.Copy(
 			coldata.CopySliceArgs{
 				SliceArgs: coldata.SliceArgs{
@@ -564,14 +560,11 @@ func (hj *hashJoiner) ExportBuffered(input Operator) coldata.Batch {
 
 func (hj *hashJoiner) resetOutput() {
 	if hj.output == nil {
-		var outColTypes []coltypes.T
-		for _, probeOutCol := range hj.spec.left.outCols {
-			outColTypes = append(outColTypes, hj.spec.left.sourceTypes[probeOutCol])
+		outputTypes := append([]coltypes.T{}, hj.spec.left.sourceTypes...)
+		if hj.spec.joinType != sqlbase.LeftSemiJoin && hj.spec.joinType != sqlbase.LeftAntiJoin {
+			outputTypes = append(outputTypes, hj.spec.right.sourceTypes...)
 		}
-		for _, buildOutCol := range hj.spec.right.outCols {
-			outColTypes = append(outColTypes, hj.spec.right.sourceTypes[buildOutCol])
-		}
-		hj.output = hj.allocator.NewMemBatch(outColTypes)
+		hj.output = hj.allocator.NewMemBatch(outputTypes)
 	} else {
 		hj.output.ResetInternalBatch()
 	}
@@ -613,16 +606,6 @@ func makeHashJoinerSpec(
 		spec                  hashJoinerSpec
 		leftOuter, rightOuter bool
 	)
-	// TODO(yuzefovich): get rid of "outCols" entirely and plumb the assumption
-	// of outputting all columns into the hash joiner itself.
-	leftOutCols := make([]uint32, len(leftTypes))
-	for i := range leftOutCols {
-		leftOutCols[i] = uint32(i)
-	}
-	rightOutCols := make([]uint32, len(rightTypes))
-	for i := range rightOutCols {
-		rightOutCols[i] = uint32(i)
-	}
 	switch joinType {
 	case sqlbase.JoinType_INNER:
 	case sqlbase.JoinType_RIGHT_OUTER:
@@ -641,22 +624,18 @@ func makeHashJoinerSpec(
 		// with the row on the right to emit it. However, we don't support ON
 		// conditions just yet. When we do, we'll have a separate case for that.
 		rightDistinct = true
-		rightOutCols = rightOutCols[:0]
 	case sqlbase.JoinType_LEFT_ANTI:
-		rightOutCols = rightOutCols[:0]
 	default:
 		return spec, errors.Errorf("hash join of type %s not supported", joinType)
 	}
 
 	left := hashJoinerSourceSpec{
 		eqCols:      leftEqCols,
-		outCols:     leftOutCols,
 		sourceTypes: leftTypes,
 		outer:       leftOuter,
 	}
 	right := hashJoinerSourceSpec{
 		eqCols:      rightEqCols,
-		outCols:     rightOutCols,
 		sourceTypes: rightTypes,
 		outer:       rightOuter,
 	}

--- a/pkg/sql/colexec/hashtable.go
+++ b/pkg/sql/colexec/hashtable.go
@@ -132,17 +132,10 @@ type hashTable struct {
 	vals coldata.Batch
 	// valTypes stores the corresponding types of the val columns.
 	valTypes []coltypes.T
-	// valCols stores the union of the keyCols and outCols.
-	valCols []uint32
 	// keyCols stores the corresponding types of key columns.
 	keyTypes []coltypes.T
 	// keyCols stores the indices of vals which are key columns.
 	keyCols []uint32
-
-	// outCols stores the indices of vals which are output columns.
-	outCols []uint32
-	// outTypes stores the types of the output columns.
-	outTypes []coltypes.T
 
 	// numBuckets returns the number of buckets the hashTable employs. This is
 	// equivalent to the size of first.
@@ -166,57 +159,13 @@ func newHashTable(
 	numBuckets uint64,
 	sourceTypes []coltypes.T,
 	eqCols []uint32,
-	outCols []uint32,
 	allowNullEquality bool,
 	mode hashTableMode,
 ) *hashTable {
-	// Compute the union of eqCols and outCols and compress vals to only keep the
-	// important columns.
-	nCols := len(sourceTypes)
-	keepCol := make([]bool, nCols)
-	compressed := make([]uint32, nCols)
-	for _, colIdx := range eqCols {
-		keepCol[colIdx] = true
-	}
-
-	for _, colIdx := range outCols {
-		keepCol[colIdx] = true
-	}
-
-	// Extract the important columns and discard the rest.
-	nKeep := uint32(0)
-
-	keepTypes := make([]coltypes.T, 0, nCols)
-	keepCols := make([]uint32, 0, nCols)
-
-	for i := 0; i < nCols; i++ {
-		if keepCol[i] {
-			keepTypes = append(keepTypes, sourceTypes[i])
-			keepCols = append(keepCols, uint32(i))
-
-			compressed[i] = nKeep
-			nKeep++
-		}
-	}
-
-	// Extract and types and indices of the	eqCols and outCols.
-	nKeys := len(eqCols)
-	keyTypes := make([]coltypes.T, nKeys)
-	keys := make([]uint32, nKeys)
-
+	keyTypes := make([]coltypes.T, len(eqCols))
 	for i, colIdx := range eqCols {
 		keyTypes[i] = sourceTypes[colIdx]
-		keys[i] = compressed[colIdx]
 	}
-
-	nOutCols := len(outCols)
-	outTypes := make([]coltypes.T, nOutCols)
-	outs := make([]uint32, nOutCols)
-	for i, colIdx := range outCols {
-		outTypes[i] = sourceTypes[colIdx]
-		outs[i] = compressed[colIdx]
-	}
-
 	ht := &hashTable{
 		allocator: allocator,
 
@@ -233,13 +182,10 @@ func newHashTable(
 			differs: make([]bool, coldata.BatchSize()),
 		},
 
-		vals:     allocator.NewMemBatchWithSize(keepTypes, 0 /* initialSize */),
-		valTypes: keepTypes,
-		valCols:  keepCols,
+		vals:     allocator.NewMemBatchWithSize(sourceTypes, 0 /* initialSize */),
+		valTypes: sourceTypes,
+		keyCols:  eqCols,
 		keyTypes: keyTypes,
-		keyCols:  keys,
-		outCols:  outs,
-		outTypes: outTypes,
 
 		numBuckets: numBuckets,
 
@@ -310,11 +256,11 @@ func (ht *hashTable) build(ctx context.Context, input Operator) {
 			}
 
 			ht.allocator.PerformOperation(targetVecs, func() {
-				for i, colIdx := range ht.valCols {
+				for i, typ := range ht.valTypes {
 					targetVecs[i].Append(
 						coldata.SliceArgs{
-							ColType:   ht.valTypes[i],
-							Src:       srcVecs[colIdx],
+							ColType:   typ,
+							Src:       srcVecs[i],
 							Sel:       batch.Selection(),
 							DestIdx:   ht.vals.Length(),
 							SrcEndIdx: batch.Length(),
@@ -404,11 +350,11 @@ func (ht *hashTable) loadBatch(batch coldata.Batch) {
 	batchSize := batch.Length()
 	ht.allocator.PerformOperation(ht.vals.ColVecs(), func() {
 		htSize := ht.vals.Length()
-		for i, colIdx := range ht.valCols {
+		for i, typ := range ht.valTypes {
 			ht.vals.ColVec(i).Append(
 				coldata.SliceArgs{
-					ColType:   ht.valTypes[i],
-					Src:       batch.ColVec(int(colIdx)),
+					ColType:   typ,
+					Src:       batch.ColVec(i),
 					Sel:       batch.Selection(),
 					DestIdx:   htSize,
 					SrcEndIdx: batchSize,

--- a/pkg/sql/colexec/mergejoiner.go
+++ b/pkg/sql/colexec/mergejoiner.go
@@ -178,10 +178,6 @@ type mergeJoinInput struct {
 	// merge join.
 	eqCols []uint32
 
-	// outCols specify the indices of the columns that should be outputted by the
-	// merge joiner.
-	outCols []uint32
-
 	// directions specifies the ordering direction of each column. Note that each
 	// direction corresponds to an equality column at the same location, i.e. the
 	// direction of eqCols[x] is encoded at directions[x], or
@@ -235,23 +231,9 @@ func newMergeJoinOp(
 	leftOrdering []execinfrapb.Ordering_Column,
 	rightOrdering []execinfrapb.Ordering_Column,
 ) (resettableOperator, error) {
-	// TODO(yuzefovich): get rid off "outCols" entirely and plumb the assumption
-	// of outputting all columns into the merge joiner itself.
-	leftOutCols := make([]uint32, len(leftTypes))
-	for i := range leftOutCols {
-		leftOutCols[i] = uint32(i)
-	}
-	rightOutCols := make([]uint32, len(rightTypes))
-	for i := range rightOutCols {
-		rightOutCols[i] = uint32(i)
-	}
-	if joinType == sqlbase.JoinType_LEFT_SEMI || joinType == sqlbase.JoinType_LEFT_ANTI {
-		rightOutCols = rightOutCols[:0]
-	}
 	base, err := newMergeJoinBase(
-		unlimitedAllocator, memoryLimit, diskQueueCfg, fdSemaphore,
-		left, right, leftOutCols, rightOutCols,
-		leftTypes, rightTypes, leftOrdering, rightOrdering,
+		unlimitedAllocator, memoryLimit, diskQueueCfg, fdSemaphore, joinType,
+		left, right, leftTypes, rightTypes, leftOrdering, rightOrdering,
 	)
 	switch joinType {
 	case sqlbase.JoinType_INNER:
@@ -305,10 +287,9 @@ func newMergeJoinBase(
 	memoryLimit int64,
 	diskQueueCfg colcontainer.DiskQueueCfg,
 	fdSemaphore semaphore.Semaphore,
+	joinType sqlbase.JoinType,
 	left Operator,
 	right Operator,
-	leftOutCols []uint32,
-	rightOutCols []uint32,
 	leftTypes []coltypes.T,
 	rightTypes []coltypes.T,
 	leftOrdering []execinfrapb.Ordering_Column,
@@ -336,16 +317,15 @@ func newMergeJoinBase(
 		memoryLimit:        memoryLimit,
 		diskQueueCfg:       diskQueueCfg,
 		fdSemaphore:        fdSemaphore,
+		joinType:           joinType,
 		left: mergeJoinInput{
 			source:      left,
-			outCols:     leftOutCols,
 			sourceTypes: leftTypes,
 			eqCols:      lEqCols,
 			directions:  lDirections,
 		},
 		right: mergeJoinInput{
 			source:      right,
-			outCols:     rightOutCols,
 			sourceTypes: rightTypes,
 			eqCols:      rEqCols,
 			directions:  rDirections,
@@ -376,6 +356,7 @@ type mergeJoinBase struct {
 	memoryLimit        int64
 	diskQueueCfg       colcontainer.DiskQueueCfg
 	fdSemaphore        semaphore.Semaphore
+	joinType           sqlbase.JoinType
 	left               mergeJoinInput
 	right              mergeJoinInput
 
@@ -426,17 +407,6 @@ func (o *mergeJoinBase) reset() {
 	o.resetBuilderCrossProductState()
 }
 
-func (o *mergeJoinBase) getOutColTypes() []coltypes.T {
-	outColTypes := make([]coltypes.T, 0, len(o.left.outCols)+len(o.right.outCols))
-	for _, leftOutCol := range o.left.outCols {
-		outColTypes = append(outColTypes, o.left.sourceTypes[leftOutCol])
-	}
-	for _, rightOutCol := range o.right.outCols {
-		outColTypes = append(outColTypes, o.right.sourceTypes[rightOutCol])
-	}
-	return outColTypes
-}
-
 func (o *mergeJoinBase) InternalMemoryUsage() int {
 	const sizeOfGroup = int(unsafe.Sizeof(group{}))
 	return 8 * coldata.BatchSize() * sizeOfGroup // o.groups
@@ -447,7 +417,11 @@ func (o *mergeJoinBase) Init() {
 }
 
 func (o *mergeJoinBase) initWithOutputBatchSize(outBatchSize int) {
-	o.output = o.unlimitedAllocator.NewMemBatchWithSize(o.getOutColTypes(), outBatchSize)
+	outputTypes := append([]coltypes.T{}, o.left.sourceTypes...)
+	if o.joinType != sqlbase.LeftSemiJoin && o.joinType != sqlbase.LeftAntiJoin {
+		outputTypes = append(outputTypes, o.right.sourceTypes...)
+	}
+	o.output = o.unlimitedAllocator.NewMemBatchWithSize(outputTypes, outBatchSize)
 	o.left.source.Init()
 	o.right.source.Init()
 	o.outputBatchSize = outBatchSize

--- a/pkg/sql/colexec/mergejoiner_test.go
+++ b/pkg/sql/colexec/mergejoiner_test.go
@@ -1978,9 +1978,8 @@ func BenchmarkMergeJoiner(b *testing.B) {
 
 				benchMemAccount.Clear(ctx)
 				base, err := newMergeJoinBase(
-					NewAllocator(ctx, &benchMemAccount), defaultMemoryLimit, queueCfg,
-					NewTestingSemaphore(mjFDLimit), leftSource, rightSource,
-					[]uint32{0, 1}, []uint32{2, 3}, sourceTypes, sourceTypes,
+					NewAllocator(ctx, &benchMemAccount), defaultMemoryLimit, queueCfg, NewTestingSemaphore(mjFDLimit),
+					sqlbase.InnerJoin, leftSource, rightSource, sourceTypes, sourceTypes,
 					[]execinfrapb.Ordering_Column{{ColIdx: 0, Direction: execinfrapb.Ordering_Column_ASC}},
 					[]execinfrapb.Ordering_Column{{ColIdx: 0, Direction: execinfrapb.Ordering_Column_ASC}},
 				)
@@ -2009,9 +2008,8 @@ func BenchmarkMergeJoiner(b *testing.B) {
 
 				benchMemAccount.Clear(ctx)
 				base, err := newMergeJoinBase(
-					NewAllocator(ctx, &benchMemAccount), defaultMemoryLimit, queueCfg,
-					NewTestingSemaphore(mjFDLimit), leftSource, rightSource,
-					[]uint32{0, 1}, []uint32{2, 3}, sourceTypes, sourceTypes,
+					NewAllocator(ctx, &benchMemAccount), defaultMemoryLimit, queueCfg, NewTestingSemaphore(mjFDLimit),
+					sqlbase.InnerJoin, leftSource, rightSource, sourceTypes, sourceTypes,
 					[]execinfrapb.Ordering_Column{{ColIdx: 0, Direction: execinfrapb.Ordering_Column_ASC}},
 					[]execinfrapb.Ordering_Column{{ColIdx: 0, Direction: execinfrapb.Ordering_Column_ASC}},
 				)
@@ -2042,9 +2040,8 @@ func BenchmarkMergeJoiner(b *testing.B) {
 
 				benchMemAccount.Clear(ctx)
 				base, err := newMergeJoinBase(
-					NewAllocator(ctx, &benchMemAccount), defaultMemoryLimit, queueCfg,
-					NewTestingSemaphore(mjFDLimit), leftSource, rightSource,
-					[]uint32{0, 1}, []uint32{2, 3}, sourceTypes, sourceTypes,
+					NewAllocator(ctx, &benchMemAccount), defaultMemoryLimit, queueCfg, NewTestingSemaphore(mjFDLimit),
+					sqlbase.InnerJoin, leftSource, rightSource, sourceTypes, sourceTypes,
 					[]execinfrapb.Ordering_Column{{ColIdx: 0, Direction: execinfrapb.Ordering_Column_ASC}},
 					[]execinfrapb.Ordering_Column{{ColIdx: 0, Direction: execinfrapb.Ordering_Column_ASC}},
 				)

--- a/pkg/sql/colexec/mergejoiner_tmpl.go
+++ b/pkg/sql/colexec/mergejoiner_tmpl.go
@@ -696,7 +696,7 @@ func _LEFT_SWITCH(_JOIN_TYPE joinTypeInfo, _HAS_SELECTION bool, _HAS_NULLS bool)
 					// We didn't materialize all the rows in the group so save state and
 					// move to the next column.
 					o.builderState.left.numRepeatsIdx += toAppend
-					if outColIdx == len(input.outCols)-1 {
+					if colIdx == len(input.sourceTypes)-1 {
 						return
 					}
 					o.builderState.left.setBuilderColumnState(initialBuilderState)
@@ -740,24 +740,23 @@ func _LEFT_SWITCH(_JOIN_TYPE joinTypeInfo, _HAS_SELECTION bool, _HAS_NULLS bool)
 // a whole.
 // SIDE EFFECTS: writes into o.output.
 func (o *mergeJoin_JOIN_TYPE_STRINGOp) buildLeftGroupsFromBatch(
-	leftGroups []group, colOffset int, input *mergeJoinInput, batch coldata.Batch, destStartIdx int,
+	leftGroups []group, input *mergeJoinInput, batch coldata.Batch, destStartIdx int,
 ) {
 	sel := batch.Selection()
 	initialBuilderState := o.builderState.left
 	outputBatchSize := o.outputBatchSize
 	o.unlimitedAllocator.PerformOperation(
-		o.output.ColVecs()[colOffset:colOffset+len(input.outCols)],
+		o.output.ColVecs()[:len(input.sourceTypes)],
 		func() {
 			// Loop over every column.
 		LeftColLoop:
-			for outColIdx, inColIdx := range input.outCols {
+			for colIdx, colType := range input.sourceTypes {
 				outStartIdx := destStartIdx
-				out := o.output.ColVec(outColIdx)
+				out := o.output.ColVec(colIdx)
 				var src coldata.Vec
 				if batch.Length() > 0 {
-					src = batch.ColVec(int(inColIdx))
+					src = batch.ColVec(colIdx)
 				}
-				colType := input.sourceTypes[inColIdx]
 
 				if sel != nil {
 					if src != nil && src.MaybeHasNulls() {
@@ -788,11 +787,7 @@ func (o *mergeJoin_JOIN_TYPE_STRINGOp) buildLeftGroupsFromBatch(
 // look at rowStartIdx and rowEndIdx). Also, all rows in the buffered group do
 // have a match, so the group can neither be "nullGroup" nor "unmatched".
 func (o *mergeJoin_JOIN_TYPE_STRINGOp) buildLeftBufferedGroup(
-	leftGroup group,
-	colOffset int,
-	input *mergeJoinInput,
-	bufferedGroup mjBufferedGroup,
-	destStartIdx int,
+	leftGroup group, input *mergeJoinInput, bufferedGroup mjBufferedGroup, destStartIdx int,
 ) {
 	var err error
 	currentBatch := o.builderState.lBufferedGroupBatch
@@ -806,18 +801,17 @@ func (o *mergeJoin_JOIN_TYPE_STRINGOp) buildLeftBufferedGroup(
 	}
 	initialBuilderState := o.builderState.left
 	o.unlimitedAllocator.PerformOperation(
-		o.output.ColVecs()[colOffset:colOffset+len(input.outCols)],
+		o.output.ColVecs()[:len(input.sourceTypes)],
 		func() {
 			batchLength := currentBatch.Length()
 			for batchLength > 0 {
 				var updatedDestStartIdx int
 				// Loop over every column.
 			LeftColLoop:
-				for outColIdx, inColIdx := range input.outCols {
+				for colIdx, colType := range input.sourceTypes {
 					outStartIdx := destStartIdx
-					src := currentBatch.ColVec(int(inColIdx))
-					out := o.output.ColVec(outColIdx)
-					colType := input.sourceTypes[inColIdx]
+					src := currentBatch.ColVec(colIdx)
+					out := o.output.ColVec(colIdx)
 					switch colType {
 					// {{ range $.MJOverloads }}
 					case _TYPES_T:
@@ -852,7 +846,7 @@ func (o *mergeJoin_JOIN_TYPE_STRINGOp) buildLeftBufferedGroup(
 							if toAppend < repeatsLeft {
 								// We didn't materialize all the rows in the current batch, so
 								// we move to the next column.
-								if outColIdx == len(input.outCols)-1 {
+								if colIdx == len(input.sourceTypes)-1 {
 									// This is the last column, so we update the builder state
 									// and exit.
 									o.builderState.left.numRepeatsIdx += toAppend
@@ -983,7 +977,7 @@ func _RIGHT_SWITCH(_JOIN_TYPE joinTypeInfo, _HAS_SELECTION bool, _HAS_NULLS bool
 				// done with the current column.
 				if toAppend < rightGroup.rowEndIdx-o.builderState.right.curSrcStartIdx {
 					// If it's the last column, save state and return.
-					if outColIdx == len(input.outCols)-1 {
+					if colIdx == len(input.sourceTypes)-1 {
 						o.builderState.right.curSrcStartIdx += toAppend
 						return
 					}
@@ -1033,18 +1027,17 @@ func (o *mergeJoin_JOIN_TYPE_STRINGOp) buildRightGroupsFromBatch(
 	outputBatchSize := o.outputBatchSize
 
 	o.unlimitedAllocator.PerformOperation(
-		o.output.ColVecs()[colOffset:colOffset+len(input.outCols)],
+		o.output.ColVecs()[colOffset:colOffset+len(input.sourceTypes)],
 		func() {
 			// Loop over every column.
 		RightColLoop:
-			for outColIdx, inColIdx := range input.outCols {
+			for colIdx, colType := range input.sourceTypes {
 				outStartIdx := destStartIdx
-				out := o.output.ColVec(outColIdx + colOffset)
+				out := o.output.ColVec(colIdx + colOffset)
 				var src coldata.Vec
 				if batch.Length() > 0 {
-					src = batch.ColVec(int(inColIdx))
+					src = batch.ColVec(colIdx)
 				}
-				colType := input.sourceTypes[inColIdx]
 
 				if sel != nil {
 					if src != nil && src.MaybeHasNulls() {
@@ -1083,7 +1076,7 @@ func (o *mergeJoin_JOIN_TYPE_STRINGOp) buildRightBufferedGroup(
 ) {
 	var err error
 	o.unlimitedAllocator.PerformOperation(
-		o.output.ColVecs()[colOffset:colOffset+len(input.outCols)],
+		o.output.ColVecs()[colOffset:colOffset+len(input.sourceTypes)],
 		func() {
 			outStartIdx := destStartIdx
 			// Repeat the buffered group numRepeats times.
@@ -1105,10 +1098,9 @@ func (o *mergeJoin_JOIN_TYPE_STRINGOp) buildRightBufferedGroup(
 					}
 
 					// Loop over every column.
-					for outColIdx, inColIdx := range input.outCols {
-						out := o.output.ColVec(outColIdx + colOffset)
-						src := currentBatch.ColVec(int(inColIdx))
-						colType := input.sourceTypes[inColIdx]
+					for colIdx, colType := range input.sourceTypes {
+						out := o.output.ColVec(colIdx + colOffset)
+						src := currentBatch.ColVec(colIdx)
 						switch colType {
 						// {{range $.MJOverloads }}
 						case _TYPES_T:
@@ -1365,14 +1357,14 @@ func (o *mergeJoin_JOIN_TYPE_STRINGOp) build() {
 		// something to build).
 		switch o.builderState.buildFrom {
 		case mjBuildFromBatch:
-			o.buildLeftGroupsFromBatch(o.builderState.lGroups, 0 /* colOffset */, &o.left, o.proberState.lBatch, outStartIdx)
+			o.buildLeftGroupsFromBatch(o.builderState.lGroups, &o.left, o.proberState.lBatch, outStartIdx)
 			// {{ if not (or _JOIN_TYPE.IsLeftSemi _JOIN_TYPE.IsLeftAnti) }}
-			o.buildRightGroupsFromBatch(o.builderState.rGroups, len(o.left.outCols), &o.right, o.proberState.rBatch, outStartIdx)
+			o.buildRightGroupsFromBatch(o.builderState.rGroups, len(o.left.sourceTypes), &o.right, o.proberState.rBatch, outStartIdx)
 		// {{ end }}
 		case mjBuildFromBufferedGroup:
-			o.buildLeftBufferedGroup(o.builderState.lGroups[0], 0 /* colOffset */, &o.left, o.proberState.lBufferedGroup, outStartIdx)
+			o.buildLeftBufferedGroup(o.builderState.lGroups[0], &o.left, o.proberState.lBufferedGroup, outStartIdx)
 			// {{ if not (or _JOIN_TYPE.IsLeftSemi _JOIN_TYPE.IsLeftAnti) }}
-			o.buildRightBufferedGroup(o.builderState.rGroups[0], len(o.left.outCols), &o.right, o.proberState.rBufferedGroup, outStartIdx)
+			o.buildRightBufferedGroup(o.builderState.rGroups[0], len(o.left.sourceTypes), &o.right, o.proberState.rBufferedGroup, outStartIdx)
 		// {{ end }}
 
 		default:

--- a/pkg/sql/colexec/unordered_distinct.go
+++ b/pkg/sql/colexec/unordered_distinct.go
@@ -28,16 +28,11 @@ func NewUnorderedDistinct(
 	colTypes []coltypes.T,
 	numHashBuckets uint64,
 ) Operator {
-	outCols := make([]uint32, len(colTypes))
-	for i := range outCols {
-		outCols[i] = uint32(i)
-	}
 	ht := newHashTable(
 		allocator,
 		numHashBuckets,
 		colTypes,
 		distinctCols,
-		outCols,
 		true, /* allowNullEquality */
 		hashTableDistinctMode,
 	)
@@ -46,7 +41,7 @@ func NewUnorderedDistinct(
 		OneInputNode: NewOneInputNode(input),
 		allocator:    allocator,
 		ht:           ht,
-		output:       allocator.NewMemBatch(ht.outTypes),
+		output:       allocator.NewMemBatch(colTypes),
 	}
 }
 
@@ -99,13 +94,13 @@ func (op *unorderedDistinct) Next(ctx context.Context) coldata.Batch {
 	nSelected = batchEnd - op.outputBatchStart
 
 	op.allocator.PerformOperation(op.output.ColVecs(), func() {
-		for i, colIdx := range op.ht.outCols {
-			toCol := op.output.ColVec(i)
-			fromCol := op.ht.vals.ColVec(int(colIdx))
+		for colIdx, typ := range op.ht.valTypes {
+			toCol := op.output.ColVec(colIdx)
+			fromCol := op.ht.vals.ColVec(colIdx)
 			toCol.Copy(
 				coldata.CopySliceArgs{
 					SliceArgs: coldata.SliceArgs{
-						ColType:     op.ht.valTypes[op.ht.outCols[i]],
+						ColType:     typ,
 						Src:         fromCol,
 						SrcStartIdx: op.outputBatchStart,
 						SrcEndIdx:   batchEnd,

